spec/text-editor-spec.coffee

```diff
@@ -4018,3 +4018,71 @@ describe "TextEditor", ->
       editor.setPlaceholderText('OK')
       expect(handler).toHaveBeenCalledWith 'OK'
       expect(editor.getPlaceholderText()).toBe 'OK'
+
+  describe 'Gutters', ->
+    describe 'Add/Remove Gutters', ->
+      it 'can add a right gutter', ->
+        options = priority: 1
+        gutter = editor.addRightGutter 'test-gutter', options
+        expect(editor.getRightGutters().length).toBe 1
+        expect(editor.getRightGutters()[0]).toBe gutter
+
+      it 'can add a left gutter', ->
+        gutter = editor.addLeftGutter 'test-gutter'
+        expect(editor.getLeftGutters().length).toBe 1
+        expect(editor.getLeftGutters()[0]).toBe gutter
+
+      it 'will notify listeners when a gutter is removed', ->
+        mockListener = onGutterRemoved: null
+        editor.onDidRemoveGutter (gutter) ->
+          mockListener.onGutterRemoved(gutter)
+        spyOn mockListener, 'onGutterRemoved'
+
+        gutter = editor.addLeftGutter 'test-gutter'
+        gutter.destroy()
+        expect(mockListener.onGutterRemoved).toHaveBeenCalledWith(gutter)
+
+    describe 'Show/Hide Gutters', ->
+      mockListener =
+      onGutterHidden: null
+      onGutterShown: null
+
+      beforeEach ->
+        editor.onDidShowGutter (gutter) ->
+          mockListener.onGutterShown(gutter)
+        editor.onDidHideGutter (gutter) ->
+          mockListener.onGutterHidden(gutter)
+        spyOn mockListener, 'onGutterShown'
+        spyOn mockListener, 'onGutterHidden'
+
+      it 'will notify listeners when a gutter is shown', ->
+        options = visibleInitially: false
+        gutter = editor.addRightGutter 'test-gutter', options
+        gutter.show()
+        expect(mockListener.onGutterShown).toHaveBeenCalledWith(gutter)
+
+      it 'will notify listeners when a gutter is hidden', ->
+        options = visibleInitially: true
+        gutter = editor.addRightGutter 'test-gutter', options
+        gutter.hide()
+        expect(mockListener.onGutterHidden).toHaveBeenCalledWith(gutter)
+
+    describe 'Default Gutter', ->
+      it 'has a non-nil default gutter', ->
+        expect(editor.getDefaultGutter()).not.toBeNull()
+
+    describe 'Decorating Gutters', ->
+      it 'adds a decoration when one of its gutters has decorateMarker called on it', ->
```

Comment:
> I think this relates to the height discussion. I think it should depend on the location of the marker. So if the marker starts at [0, 60] in buffer coordinates and the first line soft-wraps at [0, 50], we'd render the item on row 1 instead of row 0. We can also assign the height based on the number of screen lines spanned by the marker.

I guess soft-wrapping is one manifestation of the problem I was thinking of, which is if the marker spans multiple lines. E.g. [0,0] to [2,0]. Is there one item per "line"?

> Yeah... I think that works. That makes it convenient to just address decorations on a single gutter. You can create and get them. Let's see how it would be for gutters to keep their own decorations, at least as far as the public API is concerned. What do you think?

Sounds good!